### PR TITLE
[FLINK-26551][table] Flip legacy casting to disable by default

### PR DESCRIPTION
**Pull request by slinkydeveloper**
## What is the purpose of the change

This PR flips the legacy cast behavior to DISABLED by default. For details about the discussion, see https://lists.apache.org/thread/r13y3plwwyg3sngh8cz47flogq621txv.

It includes a couple of test fixes for tests relying on the bad behavior.

### CODE REVIEW

File: flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/table/KafkaTableITCase.java [hunk 1]
```diff
@@ -299,29 +301,20 @@ public void testKafkaSourceSinkWithMetadata() throws Exception {
...
         tEnv.executeSql(createTable);
 
         String initialValues =
                 "INSERT INTO kafka\n"
                         + "VALUES\n"
-                        + " ('data 1', 1, TIMESTAMP '2020-03-08 13:12:11.123', MAP['k1', X'C0FFEE', 'k2', X'BABE'], TRUE),\n"
+                        + " ('data 1', 1, TIMESTAMP '2020-03-08 13:12:11.123', MAP['k1', x'C0FFEE', 'k2', x'BABE01'], TRUE),\n"
                         + " ('data 2', 2, TIMESTAMP '2020-03-09 13:12:11.123', CAST(NULL AS MAP<STRING, BYTES>), FALSE),\n"
-                        + " ('data 3', 3, TIMESTAMP '2020-03-10 13:12:11.123', MAP['k1', X'10', 'k2', X'20'], TRUE)";
+                        + " ('data 3', 3, TIMESTAMP '2020-03-10 13:12:11.123', MAP['k1', X'102030', 'k2', X'203040'], TRUE)";
```
Comment (slinkydeveloper) [inline flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/table/KafkaTableITCase.java:311]: For more details about this, check https://issues.apache.org/jira/browse/FLINK-26549. The legacy behavior hide this issue, as no trimming and padding is performed.

File: flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/table/KafkaTableITCase.java [hunk 2]
```diff
@@ -299,29 +301,20 @@ public void testKafkaSourceSinkWithMetadata() throws Exception {
...
                                 + ")",
                         topic, bootstraps, groupId, formatOptions());
-
         tEnv.executeSql(createTable);
 
         String initialValues =
                 "INSERT INTO kafka\n"
                         + "VALUES\n"
-                        + " ('data 1', 1, TIMESTAMP '2020-03-08 13:12:11.123', MAP['k1', X'C0FFEE', 'k2', X'BABE'], TRUE),\n"
+                        + " ('data 1', 1, TIMESTAMP '2020-03-08 13:12:11.123', MAP['k1', x'C0FFEE', 'k2', x'BABE01'], TRUE),\n"
```
Comment (matriv) [inline flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/table/KafkaTableITCase.java:309]: This is not needed, please revert.

**Review by matriv — APPROVED** [review]
LGTM, left a minor comment
Please don't forget to adjust the docs, once we merge this: https://github.com/apache/flink/pull/18813

File: flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/table/KafkaTableITCase.java (hunk 2 above)
Comment (slinkydeveloper) [reply flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/table/KafkaTableITCase.java:309]: the `BABE01` is unfortunately needed

**Review by matriv — APPROVED** [review]
LGTM. Thx @slinkydeveloper!

**Pull request closed without merging**